### PR TITLE
[IMP] merge_records: add 'merge' option for many2one_reference

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Firstly I thought on using new 'first_not_null' option in chars for its model_field, but later I understood that both fields are linked and thus I used the actual approach.

It's a bit corner case, and few models have many2one_reference for now, but maybe they increase in the future.

In v14, they are used in `ir.model.data`, `ir.attachment` and several mail models.

### CODE REVIEW

File: openupgradelib/openupgrade_merge_records.py [hunk 1]
```diff
@@ -421,6 +425,23 @@ def _adjust_merged_values_orm(env, model_name, record_ids, target_record_id,
             if op == 'merge':
                 if not getattr(target_record, field.name) and _list:
                     vals[field.name] = _list[0]
+        elif field.type == 'many2one_reference':
+            if field.model_field not in model._fields:
+                continue
+            model_field = [x for x in fields if x.name == field.model_field][0]
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade_merge_records.py:431]: No need to twist the thing:
```suggestion
            model_field = model._fields[field.model_field]
```

File: openupgradelib/openupgrade_merge_records.py [hunk 2]
```diff
@@ -421,6 +425,23 @@ def _adjust_merged_values_orm(env, model_name, record_ids, target_record_id,
...
+            op = op or 'merge'
+            if op == 'merge':
+                if field.model_field not in field_spec:
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade_merge_records.py:434]: It shouldn't matter if it's on field_spec or not, we should perform the operation IMO. In fact, we must remove the entry from `field_spec` if exists.

File: openupgradelib/openupgrade_merge_records.py [hunk 3]
```diff
@@ -421,6 +425,23 @@ def _adjust_merged_values_orm(env, model_name, record_ids, target_record_id,
...
+                    list_model_field = all_records.mapped(field.model_field)
+                    zip_list = [(x, y) for x, y
+                                in zip(_list, list_model_field) if x and y]
+                    if not (
+                        getattr(target_record, field.name)
+                        and getattr(target_record, field.model_field)
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade_merge_records.py:440]: We should remove this line IMO, as we are artificially restricting the update.

File: openupgradelib/openupgrade_merge_records.py [hunk 4]
```diff
@@ -421,6 +425,23 @@ def _adjust_merged_values_orm(env, model_name, record_ids, target_record_id,
...
+                    ) and zip_list:
+                        vals[field.name] = zip_list[0][0]
+                        vals[field.model_field] = zip_list[0][1]
+                        fields.remove(model_field)
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade_merge_records.py:444]: Why removing this?

File: openupgradelib/openupgrade_merge_records.py (hunk 3 above)
Comment (MiquelRForgeFlow) [reply openupgradelib/openupgrade_merge_records.py:440]: Then we need https://github.com/OCA/openupgradelib/pull/252.

Comment (MiquelRForgeFlow) [reply openupgradelib/openupgrade_merge_records.py]: The same you could say about the `gettattr` of line https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade_merge_records.py#L422, that also could be removed, right?

Comment (pedrobaeza) [reply openupgradelib/openupgrade_merge_records.py]: It's not the same. I'm talking about removing this specific line, not the previous one. The first check is for not overwriting if there's already a value set. This second one is for forcing to have a value in the other field.

Comment (pedrobaeza) [reply openupgradelib/openupgrade_merge_records.py]: Anyways, this is not critical of course. Both fields should be filled in normal conditions and not filled on contrary. To have one filled and the other not is the anomaly.

Comment (MiquelRForgeFlow) [reply openupgradelib/openupgrade_merge_records.py]: Imagine you have target record with m2o_r = 0, model = 'ir.attachment', record 2 with m2o_r = 0, model = 'ir.attachment', and record 3 with m2o_r = 1, model = 'ir.attachment'. You want target_record having m2o_r = 1, model = 'ir.attachment'. That's why we need double check.

Comment (MiquelRForgeFlow) [reply openupgradelib/openupgrade_merge_records.py]: Yeah, you are right. Done.

**Comment by MiquelRForgeFlow** [comment]
Ready

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**